vllm/model_executor/models/baichuan.py

```diff
@@ -351,8 +352,42 @@ def load_weights(self,
             ("gate_up_proj", "up_proj", 1),
         ]
         params_dict = dict(self.named_parameters())
-        for name, loaded_weight in hf_model_weights_iterator(
-                model_name_or_path, cache_dir, load_format, revision):
+        weight_names = []
+        for name in params_dict:
+            if "gate_up_proj" in name:
+                weight_names.append(name.replace("gate_up_proj", "gate_proj"))
+                weight_names.append(name.replace("gate_up_proj", "up_proj"))
+            else:
+                weight_names.append(name)
+        weight_names = sorted(weight_names)
+
+        def weights_iterator(
+            use_distributed_loading: bool
+        ) -> Iterator[Tuple[str, Optional[torch.Tensor], Optional[int]]]:
+            if use_distributed_loading:
+                local_state_dict, weights_map = \
+                    hf_model_weights_distributed_loader(
+                        model_name_or_path, weight_names,
+                        cache_dir, load_format, revision)
+                for name in weight_names:
+                    weight_owner = weights_map[name]
+                    loaded_weight = local_state_dict[name].to("cuda") \
```

Comment:
Why `"cuda"` in particular?

Reply:
Because the loaded_weight is on the cpu after loading, but the params are on the gpu, and the transfer rate of tensors on the gpu is faster.

Reply:
To clarify - I'm asking why we need to hardcode "cuda" here since vLLM is supported by multiple hardware types. If this would only work on CUDA devices then we might need to do a check at the engine initialization.

Reply:
Got it. I think this approach only works on CUDA devices right now, because the speed of transferring data across multiple ranks might not be fast enough on other devices to be worth it. I've added some checks for this at the engine initialization.